aten/src/ATen/OpaqueTensorImpl.h

```diff
@@ -77,35 +77,44 @@ struct CAFFE2_API OpaqueTensorImpl : public TensorImpl {
     AT_ERROR("opaque tensors do not have storage");
   }
 
-// NOTE: `shallow_copy_and_detach()` does not copy the following TensorImpl fields:
-// 1. the AutogradMeta pointer, because it is unique for each Variable.
-// 2. the version counter, because it is set to the passed in `version_counter`.
-//    See NOTE [ Version Counter Sharing ] for details.
-//
-// NOTE: `allow_tensor_metadata_change` determines whether the TensorImpl shallow-copy
-// allows changes to its metadata (e.g. sizes / strides / storage / storage_offset).
-// See NOTE [ Metadata Change for a Detached Tensor ] for details.
+/**
+ * Return a TensorImpl that is a shallow-copy of this TensorImpl.
+ *
+ * For usage of `version_counter` and `allow_tensor_metadata_change`,
+ * see NOTE [ TensorImpl Shallow-Copying ].
+ */
 c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
     const c10::VariableVersion& version_counter,
     bool allow_tensor_metadata_change) const override {
   //AT_ASSERT(false);
   auto impl = c10::make_intrusive<OpaqueTensorImpl<OpaqueHandle>>(
     type_id(), dtype(), device(), opaque_handle_, sizes_);
-  // TensorImpl general fields
-  // Note that some of these fields are not used in opaque tensor code,
-  // and we copy them here only for completeness.
-  impl->sizes_ = sizes_;
-  impl->strides_ = strides_;
-  impl->storage_offset_ = storage_offset_;
-  impl->is_contiguous_ = is_contiguous_;
-  impl->is_wrapped_number_ = is_wrapped_number_;
-  impl->reserved_ = reserved_;
-  impl->set_version_counter(version_counter);
-  impl->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
-
-  // OpaqueTensorImpl-specific fields (none currently).
+  copy_tensor_data(
+    /*src_impl=*/this,
+    /*dest_impl=*/impl.get(),
+    /*version_counter=*/version_counter,
+    /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+
+  // OpaqueTensorImpl-specific fields.
+  impl->opaque_handle_ = opaque_handle_;
```

Comment:
is there a way we can avoid having to specify the specific fields twice?

Can `tensor_copy_data` be virtual?

Reply:
`tensor_copy_data` is made virtual now.